File: 回溯/17.电话号码的字母组合.java
class Solution {
    List<String> res;
    public List<String> letterCombinations(String digits) {
        res = new ArrayList<>();
        if(digits == null || digits.length()==0){
            return res;
        }
        String[] numString ={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        backing(digits,numString,0);
        return res;
    }
    StringBuilder sb = new StringBuilder();
    void backing(String digits,String[] numString,int start){
        if(start == digits.length()){
            res.add(sb.toString());
            return;
        }
        String str = numString[digits.charAt(start)-'0'];
        for(int i=0;i<str.length();i++){
            sb.append(str.charAt(i));
            backing(digits,numString,start+1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 回溯/17.电话号码的字母组合.go
func letterCombinations(digits string) []string {
	n := len(digits)
	if n == 0 {
		return []string{}
	}
	m := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}
	res := make([]string, 0)
	path := make([]byte, 0)
	var dfs func(int)
	dfs = func(start int) {
		if start == len(digits) {
			res = append(res, string(path))
			return
		}
		str := m[digits[start]]
		for i := 0; i < len(str); i++ {
			path = append(path, str[i])
			dfs(start + 1)
			path = path[:len(path)-1]
		}
	}
	dfs(0)
	return res
}

